src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeNameParser.cs

```diff
@@ -188,14 +188,22 @@ private TypeNameParser(ReadOnlySpan<char> name, bool throwOnError, TypeNameParse
             if (allowFullyQualifiedName && !TryParseAssemblyName(ref assemblyName))
             {
 #if SYSTEM_PRIVATE_CORELIB
-                // backward compat: throw FileLoadException for non-empty invalid strings
-                if (_throwOnError || !_inputString.TrimStart().StartsWith(","))
+                // Backward compatibility: throw for non-empty invalid assembly names.
+                if (!_inputString.TrimStart().StartsWith(","))
                 {
-                    throw new IO.FileLoadException(SR.InvalidAssemblyName, _inputString.ToString());
+                    // No matter what throwOnError is set to, if top level assembly was not provided, CLR throws FileLoadException for invalid assembly names.
+                    if (!_parseOptions.TopLevelAssemblyWasProvided)
+                    {
+                        throw new IO.FileLoadException(SR.InvalidAssemblyName, _inputString.ToString());
+                    }
+                    // If top level was provided and user has requested for error, we throw ArgumentException.
+                    else if (_throwOnError)
+                    {
+                        throw new ArgumentException(SR.Argument_AssemblyGetTypeCannotSpecifyAssembly);
+                    }
```

Comment:
```suggestion
                    // Reject attempt to provide top-level assembly name to Assembly.GetType
                    if (_parseOptions.IsAssemblyGetType)
                    {
                        if (_throwOnError)
                        {
                            throw new ArgumentException(SR.Argument_AssemblyGetTypeCannotSpecifyAssembly);
                        }
                        return null;
                    }

                    // Otherwise, no matter what throwOnError is set to, we throw FileLoadException for invalid assembly names.
                    throw new IO.FileLoadException(SR.InvalidAssemblyName, _inputString.ToString());
```
Nit: It took me a bit of time to connect the dots between `TopLevelAssemblyWasProvided` and `Argument_AssemblyGetTypeCannotSpecifyAssembly` message. I think the code would be easier to understand if we  use `IsAssemblyGetType` for the property and otherwise keep the code to handle `Assembly.GetType` case together.

Also, unify comment style "CLR throws" vs. "we throw".